src/task/mod.rs

```diff
@@ -35,16 +36,38 @@ pub trait Task: Send + Sized + 'static {
     /// ```js
     /// function callback(err, value) {}
     /// ```
+    #[cfg(feature = "legacy-runtime")]
     fn schedule(self, callback: Handle<JsFunction>) {
-        let boxed_self = Box::new(self);
-        let self_raw = Box::into_raw(boxed_self);
-        let callback_raw = callback.to_raw();
-        unsafe {
-            neon_runtime::task::schedule(mem::transmute(self_raw),
-                                         perform_task::<Self>,
-                                         complete_task::<Self>,
-                                         callback_raw);
-        }
+        schedule_internal(Env::current(), self, callback);
+    }
+
+    /// Schedule a task to be executed on a background thread.
+    ///
+    /// `callback` should have the following signature:
+    ///
+    /// ```js
+    /// function callback(err, value) {}
+    /// ```
+    #[cfg(feature = "napi-runtime")]
+    fn schedule<'a, C: crate::context::Context<'a>>(
+        self,
+        cx: &mut C,
+        callback: Handle<JsFunction>,
+    ) {
+        schedule_internal(cx.env(), self, callback);
+    }
+}
+
+fn schedule_internal<T: Task>(env: Env, task: T, callback: Handle<JsFunction>) {
```

Comment:
This helper method is consistent across legacy and `n-api`. The breaking change to require `Context` is implemented on the trait.